sdk/storage/Azure.Storage.Queues/src/QueueClient.cs

```diff
@@ -362,44 +360,57 @@ public QueueClient(Uri queueUri, TokenCredential credential, QueueClientOptions
         /// name of the account, and the name of the queue.
         /// This is likely to be similar to "https://{account_name}.queue.core.windows.net/{queue_name}".
         /// </param>
-        /// <param name="pipeline">
-        /// The transport pipeline used to send every request.
-        /// </param>
-        /// <param name="storageSharedKeyCredential">
-        /// The shared key credential used to sign requests.
-        /// </param>
-        /// <param name="version">
-        /// The version of the service to use when sending requests.
-        /// </param>
-        /// <param name="clientDiagnostics">
-        /// The <see cref="ClientDiagnostics"/> instance used to create
-        /// diagnostic scopes every request.
-        /// </param>
-        /// <param name="encryptionOptions">
-        /// Options for client-side encryption.
-        /// </param>
-        /// <param name="messageEncoding">
-        /// The encoding of the message sent over the wire.
+        /// <param name="clientConfiguration">
+        /// <see cref="QueueClientConfiguration"/>.
         /// </param>
         internal QueueClient(
             Uri queueUri,
-            HttpPipeline pipeline,
-            StorageSharedKeyCredential storageSharedKeyCredential,
-            QueueClientOptions.ServiceVersion version,
-            ClientDiagnostics clientDiagnostics,
-            ClientSideEncryptionOptions encryptionOptions,
-            QueueMessageEncoding messageEncoding)
+            QueueClientConfiguration clientConfiguration)
         {
             _uri = queueUri;
             _messagesUri = queueUri.AppendToPath(Constants.Queue.MessagesUri);
-            _pipeline = pipeline;
-            _storageSharedKeyCredential = storageSharedKeyCredential;
-            _version = version;
-            _clientDiagnostics = clientDiagnostics;
-            _clientSideEncryption = QueueClientSideEncryptionOptions.CloneFrom(encryptionOptions);
-            _messageEncoding = messageEncoding;
+            _clientConfiguration = clientConfiguration;
+
+            (QueueRestClient, MessagesRestClient, MessageIdRestClient) clients = BuildRestClients();
+            _queueRestClient = clients.Item1;
+            _messagesRestClient = clients.Item2;
+            _messageIdRestClient = clients.Item3;
+
             AssertEncodingForEncryption();
         }
+
+        private (QueueRestClient, MessagesRestClient, MessageIdRestClient) BuildRestClients()
+        {
+            QueueUriBuilder uriBuilder = new QueueUriBuilder(_uri);
+            string queueName = uriBuilder.QueueName;
+            uriBuilder.QueueName = null;
+
+            string uriString = uriBuilder.ToUri().ToString();
```

Comment:
we're already decomposing uri in some constructors. Maybe we should take `QueueUriBuilder` as parameter here to avoid double parsing?

Reply:
We would be changing the state of the `QueueUriBuilder`, since I have to remove the queue name.  I'd be concerned someone might try to use the `QueueUriBuilder` after calling `BuildRestClients()`, causing them to get an unexpected Uri.